pkg/controller/cloneset/scale/cloneset_scale.go

```diff
@@ -0,0 +1,207 @@
+		pods []*v1.Pod, pvcs []*v1.PersistentVolumeClaim,
+	) (bool, error)
+}
+
+// NewScaleControl returns a scale control.
+func NewScaleControl(c client.Client, recorder record.EventRecorder, exp expectations.ScaleExpectations) Interface {
+	return &realControl{Client: c, recorder: recorder, exp: exp}
+}
+
+type realControl struct {
+	client.Client
+	recorder record.EventRecorder
+	exp      expectations.ScaleExpectations
+}
+
+func (r *realControl) ManageReplicas(
+	currentCS, updateCS *appsv1alpha1.CloneSet,
+	currentRevision, updateRevision string,
+	pods []*v1.Pod, pvcs []*v1.PersistentVolumeClaim,
+) (bool, error) {
+	if updateCS.Spec.Replicas == nil {
+		return false, fmt.Errorf("spec.Replicas is nil")
+	}
+
+	if podsToDelete := getPodsToDelete(updateCS, pods); len(podsToDelete) > 0 {
+		return true, r.deletePods(updateCS, podsToDelete, pvcs)
+	}
+
+	var err error
+	diff := len(pods) - int(*updateCS.Spec.Replicas)
+
+	if diff < 0 {
+		// total number of this creation
+		expectedCreations := diff * -1
+		// lack number of current version
+		expectedCurrentCreations := 0
+		if updateCS.Spec.UpdateStrategy.Partition != nil {
+			notUpdatedNum := len(pods) - len(clonesetutils.GetUpdatedPods(pods, updateRevision))
+			expectedCurrentCreations = int(*updateCS.Spec.UpdateStrategy.Partition) - notUpdatedNum
+		}
+
+		// available instance-id come from free pvc
+		availableIDs := getOrGenAvailableIDs(expectedCreations, pods, pvcs)
+		// existing pvc names
+		existingPVCNames := sets.NewString()
+		for _, pvc := range pvcs {
+			existingPVCNames.Insert(pvc.Name)
+		}
+
+		err = r.createPods(expectedCreations, expectedCurrentCreations,
+			currentCS, updateCS, currentRevision, updateRevision, availableIDs.List(), existingPVCNames)
+		return true, err
+
+	} else if diff > 0 {
+		podsToDelete := r.choosePodsToDelete(pods, diff)
+		err = r.deletePods(updateCS, podsToDelete, pvcs)
+
+		return true, err
+	}
+
+	return false, nil
+}
+
+func (r *realControl) createPods(
+	expectedCreations, expectedCurrentCreations int,
+	currentCS, updateCS *appsv1alpha1.CloneSet,
+	currentRevision, updateRevision string,
+	availableIDs []string, existingPVCNames sets.String,
+) error {
+	// list of all pods need to create
+	var newPods []*v1.Pod
+	if expectedCreations <= expectedCurrentCreations {
+		newPods = newVersionedPods(currentCS, currentRevision, expectedCreations, &availableIDs)
+	} else {
+		newPods = newVersionedPods(currentCS, currentRevision, expectedCurrentCreations, &availableIDs)
+		newPods = append(newPods, newVersionedPods(updateCS, updateRevision, expectedCreations-expectedCurrentCreations, &availableIDs)...)
+	}
+
+	podsCreationChan := make(chan *v1.Pod, len(newPods))
+	for _, p := range newPods {
+		r.exp.ExpectScale(clonesetutils.GetControllerKey(updateCS), expectations.Create, p.Name)
+		podsCreationChan <- p
+	}
+
+	successPodNames := sync.Map{}
+	_, err := clonesetutils.DoItSlowly(len(newPods), initialBatchSize, func() error {
```

Comment:
There are multiple goroutines in `DoItSlowly`, using index cloud make the logic more complicated. 
(Concurrently get unique index, ready list)

The item in channel is `*v1.Pod`, which is a pointer, it will not waste memory. The ReplicaSet controller in Kubernetes is similar to this.